src/Components/test/E2ETest/ServerExecutionTests/InteropReliabilityTests.cs

```diff
@@ -0,0 +1,349 @@
+                null,
+                JsonSerializer.Serialize(Array.Empty<object>()));
+
+            await Task.Delay(DefaultLatencyTimeout);
+
+            Assert.Single(interopCalls, (0, "DotNet.jsCallDispatcher.endInvokeDotNetFromJS", expectedImportantDotNetObjectRef));
+
+            // Act
+            Client.InvokeDotNetMethod(
+                "1",
+                "BasicTestApp",
+                "ReceiveTrivial",
+                null,
+                JsonSerializer.Serialize(new object[] { new { __dotNetObject = 1 } }));
+
+            await Task.Delay(DefaultLatencyTimeout);
+
+            // Assert
+            Assert.Single(interopCalls, (0, "DotNet.jsCallDispatcher.endInvokeDotNetFromJS", expectedError));
+
+            await ValidateClientKeepsWorking(Client, batches);
+        }
+
+        [Fact]
+        public async Task ContinuesWorkingAfterInvalidAsyncReturnCallback()
+        {
+            // Arrange
+            var expectedError = "An exception occurred executing JS interop: The JSON value could not be converted to System.Int32. Path: $ | LineNumber: 0 | BytePositionInLine: 3.. See InnerException for more details.";
+
+            var (interopCalls, batches) = ConfigureClient();
+            await GoToTestComponent(batches);
+
+            // Act
+            var jsInteropTriggered = Client.PrepareForNextBatch();
+            await Client.ClickAsync("triggerjsinterop");
+
+            await Task.Delay(DefaultLatencyTimeout);
+            Assert.Single(interopCalls, (4, "sendMalformedCallbackReturn", (string)null));
+
+            var invalidJSInteropResponse = Client.PrepareForNextBatch();
+            Client.InvokeDotNetMethod(
+                0,
+                "Microsoft.JSInterop",
+                "DotNetDispatcher.EndInvoke",
+                null,
+                "[4, true, \"{\"]");
+
+            await invalidJSInteropResponse;
+            var text = Assert.Single(
+                Client.FindElementById("errormessage").Children.OfType<TextNode>(),
+                e => expectedError == e.TextContent);
+
+            await ValidateClientKeepsWorking(Client, batches);
+        }
+
+        [Fact]
+        public async Task CannotInvokeJSInvokableMethodsWithInvalidArgumentsPayload()
+        {
+            // Arrange
+            var expectedError = "[\"1\"," +
+                "false," +
+                "\"There was an exception invoking \\u0027NotifyLocationChanged\\u0027 on assembly \\u0027Microsoft.AspNetCore.Components.Server\\u0027. For more details turn on detailed exceptions in \\u0027CircuitOptions.JSInteropDetailedErrors\\u0027\"]";
+
+            var (interopCalls, batches) = ConfigureClient();
+            await GoToTestComponent(batches);
+
+            // Act
+            Client.InvokeDotNetMethod(
+                "1",
+                "Microsoft.AspNetCore.Components.Server",
+                "NotifyLocationChanged",
+                null,
+                "[ \"invalidPayload\"}");
```

Comment:
Is it worth embedded different invalid JSON parts into the different parts of this payload? Like what if the method and assembly are null, but the args is invalid.

Reply:
They all get parsed the same way, I don't think it adds value, we can add a test that includes an invalid argument, given that we parse the array and then we parse each argument, but I don't think there's a lot of value in it.

Reply:
It seemed relevant to me given the level of detail that the other tests were going into. We're likely going to do more work in this are to use the JSON reader instead of the document. I'm happy to see this kind of test added then if you'd prefer not to do it.